osu.Server.Spectator/Hubs/MultiplayerQueue.cs

```diff
@@ -155,61 +156,106 @@ public async Task AddItem(MultiplayerPlaylistItem item, MultiplayerRoomUser user
         /// Duplicates <see cref="CurrentItem"/> into the database.
         /// </summary>
         /// <param name="db">The database connection.</param>
-        private async Task duplicateCurrentItem(IDatabaseAccess db)
+        private async Task duplicateCurrentItem(IDatabaseAccess db) => await addItem(db, new MultiplayerPlaylistItem
         {
-            var newItem = new MultiplayerPlaylistItem
-            {
-                OwnerID = CurrentItem.OwnerID,
-                BeatmapID = CurrentItem.BeatmapID,
-                BeatmapChecksum = CurrentItem.BeatmapChecksum,
-                RulesetID = CurrentItem.RulesetID,
-                AllowedMods = CurrentItem.AllowedMods,
-                RequiredMods = CurrentItem.RequiredMods
-            };
-
-            newItem.ID = await db.AddPlaylistItemAsync(new multiplayer_playlist_item(room.RoomID, newItem));
-            room.Playlist.Add(newItem);
-
-            await hub.OnPlaylistItemAdded(room, newItem);
+            OwnerID = CurrentItem.OwnerID,
+            BeatmapID = CurrentItem.BeatmapID,
+            BeatmapChecksum = CurrentItem.BeatmapChecksum,
+            RulesetID = CurrentItem.RulesetID,
+            AllowedMods = CurrentItem.AllowedMods,
+            RequiredMods = CurrentItem.RequiredMods
+        });
+
+        private async Task addItem(IDatabaseAccess db, MultiplayerPlaylistItem item)
+        {
+            // Add the item to the list first in order to compute gameplay order before adding to the database.
+            room.Playlist.Add(item);
+            await updatePlaylistOrder(db);
+
+            // Actually add the item to the database and invoke callback on the hub.
+            item.ID = await db.AddPlaylistItemAsync(new multiplayer_playlist_item(room.RoomID, item));
+            await hub.OnPlaylistItemAdded(room, item);
         }
 
         /// <summary>
         /// Updates <see cref="CurrentItem"/> and the playlist item ID stored in the room's settings.
         /// </summary>
         private async Task updateCurrentItem()
         {
-            MultiplayerPlaylistItem newItem;
+            MultiplayerPlaylistItem nextItem = room.Playlist
+                                                   .Where(i => !i.Expired)
+                                                   .OrderBy(i => i.PlaylistOrder)
+                                                   .FirstOrDefault()
+                                               ?? room.Playlist.Last();
```

Comment:
Shouldn't this fallback "last" case also be sorted by `PlaylistOrder` just for sanity?

Reply:
True, good catch.